5주차 과제1 (레스토랑 조회 구현하기)

seong-dodo opened the pull request:
안녕하세요.

5주차 과제 진행하여 차근차근 올리겠습니다.
지난주 과제에 접목하지 못하였던 피드백도 최대한 반영하여 코드 구현 하도록 해보겠습니다 🔥 🔥 🔥 🔥 🔥

seong-dodo committed:
1. App 테스트 코드 추가

- 레스토랑 지역 목록을 그린다에 대한 테스트 코드 추가하기

seong-dodo committed:
2. App 컴포넌트 기능 구현

- 레스토랑 지역 목록에 대한 기능 구현

seong-dodo committed:
- actions 파일 생성
- reducer 파일 & reducer 테스트 파일 생성
- store 파일 생성
- store 코드 작성 & Provider를 통해 redux store에 접근하도록 하기

seong-dodo committed:
1. RegionsContainer 테스트 추가 &  mocks 모듈생성

- RegionsContainer 테스트 추가
- mocks 모듈화
- reducer 테스트 파일 주석 처리

seong-dodo committed:
2. RegionsContaine 컴포넌트 기능 구현

- RegionsContaine에 지역 목록에 대한 기능 구현

seong-dodo committed:
1. Regions 테스트 코드 추가

- Regions 테스트 코드 추가
- Regions 테스트파일 & Regions 컴포넌트 파일 생성

seong-dodo committed:
1. Regions 컴포넌트 기능 구현

-  Regions에 대한 기능 구현
- 변수명 수정 (region -> name)

seong-dodo committed:
레스토랑 지역 목록 API 가져오기 위한 코드 추가

- App컴포넌트에 레스토랑 지역 목록 가져오기 위해 코드 추가

seong-dodo committed:
1. reducer 테스트 코드 추가

- Regions이 보여진다

seong-dodo committed:
2. reducer 컴포넌트 기능 구현 및 API 처리 & 오류, 오타 수정

- reducer 컴포넌트 기능 구현
- API 처리 : 레스토랑 지역 목록
- 오타 및 오류 수정

seong-dodo committed:
1. Categories 테스트 코드 추가

- Categories 테스트 파일 생성 및 테스트 코드 추가
- Categories 컴포넌트 파일 생성

seong-dodo committed:
2. Categories 컴포넌트 기능 구현

- Categories에 대한 기능 구현

seong-dodo committed:
1. App 테스트 코드 추가

레스토랑 분류 목록을 그린다에 대한 테스트 코드 추가하기

seong-dodo committed:
2. App 컴포넌트 기능 구현

- App 컴포넌트에 Categories컴포넌트 연결하기

seong-dodo committed:
1. CategoriesContainer 테스트 코드 추가

- CategoriesContainer 테스트 추가

seong-dodo committed:
2. CategoriesContainer 컴포넌트 기능 구현

- CategoriesContainer 컴포넌트와 App컴포넌트 연결

seong-dodo committed:
레스토랑 분류 목록 API 가져오기 위한 코드 추가

- App컴포넌트에 레스토랑 분류 목록 가져오기 위해 코드 추가
- API 처리 : 레스토랑 분류 목록

seong-dodo committed:
reducer 테스트 코드 추가, 기능 구현

- reducer 테스트 코드 추가
- reducer 기능 구현
- 오류수정 & App 테스트 수정

seong-dodo commented:
안녕하세요, 트레이너님
지난주차 과제에서  리듀서 if-else문을 제거하고 object literals을 적용해보는게 어떻냐고 하셔서 이번 주차 과제에 한번 적용을 해보려고 합니다.


리듀서가 초기화 되면서 아래와 같은 액션이 항상 발생하고 있는데 payload가 없는 action이길래 코드에서처럼 action.payload===undefined일경우를 if문으로 처리했습니다... 혹시 제가 적용한 object literals 에서 리덕스 초기화 액션을 어떻게 예외처리해야하는게 좋을까요

![image](https://user-images.githubusercontent.com/68280407/120496412-82481580-c3f8-11eb-861d-daf0e4e9edda.png)

```javascript
export default function reducer(state = initialState, action) {
  if (action === undefined) {
    return state;
  }
  if (action.payload === undefined) {
    return state;
  }

  return actionTable(state, action);
}

```

seong-dodo committed:
reducer if문 제거

- reducer if문 제거, object literals 사용

seong-dodo committed:
App 테스트 추가

- api mocking  테스트 추가

moonkii commented on src/App.test.jsx:
```diff
@@ -0,0 +1,26 @@
+      categories: [
+        { id: 1, name: '한식' },
+      ],
+    }));
```
리덕스 부분은 테스트 별 컴포넌트가 랜더링 될 때 마다 필요한 부분이에요.
그렇기 때문에 `it` 안에 넣어주기 보다 밖으로 빼주시는 게 좋아요.

moonkii commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,15 @@
+
+  const { getByText } = render(<Categories categories={categories} />);
+
+  expect(getByText(/한식/)).not.toBeNull();
```
`getBy~` 와 `queryBy~` 의 차이점을 아시나요?

한 번 찾아보시면 도움되실 거에요 ;)

seong-dodo replied:
제가 찾아보고 공부를 해보았는데 질문이 있어 문의 드립니다.

`getBy*`, `queryBy*` 차이점 : DOM 엘리먼트가 존재하지 않을 때, 에러 발생 여부 
- `getby*` : 에러 발생
- `queryBy*` : 에러 발생 x

왜? `queryBy*`를 써야하는가? 어떠한 경우에 사용해야 하는가?
제가 생각할 때는 정확한 테스트 코드를 만들기 위해서는 `getBy*` 를 사용해야 하는게 아닌가 하는 의문이 드네요..

moonkii replied:
`queryBy~` 는 에러 발생하지 않는다 라기 보단 엘리먼트가 존재하지 않으면 null 을 반환한다 라고 보시면 좋아요.

결과적으로는 어차피 테스트가 실패하기 때문에 동일하다고 생각할수도 있지만
의도를 드러내는 관점에서 보면 `toBeNull()` 는 true 이냐 false 이냐를 판별하기 때문에
에러를 발생시키는게 좋을지 boolean 을 반환하는게 좋을지 고민해보시면 좋을 것 같아요.

seong-dodo replied:
감사합니다 👍 
말씀해주신대로 고민해 보니  이 부분에서는 `queryBy~`를 사용해야할 것 같습니다

moonkii commented on src/reducer.js:
```diff
@@ -0,0 +1,28 @@
+  categories: [],
+};
+
+const actionTable = (state, action) => ({
```
`actionTable` 이라고하면 의미가 잘 드러나지 않는 것 같아요.

어떻게 바꿔주면 좋을까요?

seong-dodo replied:
안녕하세요.
`stateByActionType`으로 변경해보았습니다!!!

moonkii replied:
리턴하는 객체는 어떤 것들을 담고 있는지 생각해보면 이름이 더 명확해질 것 같아요!

moonkii commented on src/reducer.js:
```diff
@@ -0,0 +1,28 @@
+  }
+  if (action.payload === undefined) {
+    return state;
+  }
```
이 예외처리들이 꼭 필요한가요?

seong-dodo replied:
감사합니다. 말씀해주신대로 Default 부분 관련해서 테스트를 작성해보고 예외처리 삭제하였습니다 👍

moonkii commented on src/reducer.test.js:
```diff
@@ -0,0 +1,36 @@
+} from './actions';
+
+describe('reducer', () => {
+  it('Regions이 보여진다', () => {
```
테스트 시나리오가 `Regions이 보여진다` 라기 보다 상태가 어떻게 변하는지 서술해주면 좋을 것 같아요.

moonkii reviewed (commented):
![image](https://user-images.githubusercontent.com/46776329/120530095-18d7ff00-c418-11eb-93b1-90e4da7301a1.png)


`if-else` 는 잘 제거하신 것 같아요!
말씀하신 현상은 `Default` 부분 때문에 발생하는 것 같아요.
관련해서 테스트를 작성해보시면 어떻게 나오고 왜 안되는지 더 찾기 쉬울거에요.
이 부분을 수정해보시면 예외처리도 따로 필요하지 않을거에요!

seong-dodo committed:
1. Categories 테스트 추가

- Categies 리스트에 버튼태그가 있는지 확인하는 테스트 코드 추가

seong-dodo committed:
1. Categories 테스트 코드 수정

- Categories 테스트 코드 수정

seong-dodo committed:
2. Categories 기능 구현, 테스트 코드 수정

- Categories 기능 구현 & 테스트 코드 수정

seong-dodo committed:
1단계: reducer 테스트 코드 추가

- default action type에 대한 테스트 코드 추가

seong-dodo committed:
2단계: reducer 기능 구현 / 피드백반영

- reducer 기능 구현
- 피드백 반영

seong-dodo committed:
1단계: Regions 테스트 추가

- Regions 리스트에 버튼태그 확인하는 테스트 코드 추가

seong-dodo committed:
2단계: Regions 기능구현

- Regions 컴포넌트에 버튼 태그 만들기

seong-dodo committed:
RegionsContainer 테스트: 피드백반영

- beforeEach 사용
- 테스트 시나리오 변경
- getByText -> queryByText로 수정

seong-dodo committed:
CategoriesContainer 테스트: 피드백반영

- beforeEach 사용
- 테스트 시나리오 변경
- getByText -> queryByText로 수정

seong-dodo committed:
1단계: Button 테스트 코드 추가

- Button 컴포넌트 분리를 위한 테스트 코드 추가

seong-dodo committed:
2단계: Button컴포넌트 기능 구현 / 1단계: 테스트 코드 추가(수정)

- Button 컴포넌트 관심사 분리 기능 구현
- 테스트 코드 추가 수정

seong-dodo committed:
1단계: reducer 테스트 추가

- reducer 동작 테스트 추가

seong-dodo committed:
2단계: reducer selecteRegion 기능 구현

- reducer : selecteRegion 동작 기능 구현
- selecteRegion actions 코드 추가
- 리듀서 테스트 코드 수정 추가 : selecteRegion

moonkii commented on src/Button.jsx:
```diff
@@ -0,0 +1,19 @@
+  function handleClick() {
+    return {
+    };
+  }
```
이부분은 뭔가요??

만약 구현이 완성되지 않은 상태라면 `// TODO: ...` 같이 주석을 달아주세요.
이런경우에 사용하는게 주석이에요 :)

moonkii commented on src/Button.test.jsx:
```diff
@@ -0,0 +1,14 @@
+  const { queryByText } = render(<Button item={item} />);
+  const button = queryByText('서울');
+
+  fireEvent.click(queryByText('서울'));
```
여기서는 클릭하는 대상이 없다면 `null` 인게 좋을까요 아니면 에러를 발생시키는게 좋을까요?

언제 `getBy~` 와 `queryBy~` 를 사용하면 좋을지 고민해보시면 좋아요!

seong-dodo replied:
여기서는 클릭하는 대상이 없으면 에러 발생 시키는게 더 나은 것 같습니다.
`getBy~`로 수정이 필요할 것 같습니다.

moonkii commented on src/reducer.test.js:
```diff
@@ -0,0 +1,64 @@
+
+describe('reducer', () => {
+  describe('setRegions', () => {
+    it('get regions object and changes regions', () => {
```
테스트 시나리오를 작성하실 떄 `it`은 3인칭 단수이기 때문에 동사에 항상 `s` 를 붙이는걸 까먹지마세요!

seong-dodo replied:
넵 갑사합니다 😄  수정하겠습니다~~~

seong-dodo committed:
1단계 reducer selecteCategory 테스트 코드 추가

- 리듀서 테스트 코드 추가: gets selecteCategory object

seong-dodo committed:
2단계:selecteCategory 기능 구현

- reducer : selecteCategory 동작 기능 구현
- selecteCategory actions 코드 추가
- actions selectRegion 매개 변수 이름 변경: item -> region

moonkii commented on src/Categories.jsx:
```diff
@@ -0,0 +1,17 @@
+    <ul>
+      {categories.map((category) => (
+        <li key={category.id}>
+          <Button
+            item={category}
+          />
```
컴포넌트를 사용하는 쪽에서 인터페이스를 미리 정의해주면 좋아요.
여기서는 클릭 이벤트 핸들러 함수가 `props` 로 들어가겠죠?

moonkii commented on src/actions.js:
```diff
@@ -12,9 +12,16 @@ export function setCategories(categories) {
 }
 
-export function selecteRegion(item) {
+export function selecteRegion(region) {
```
매개변수 이름을 명확하게 해주니 가독성이 더 좋은 것 같아요 👍

seong-dodo committed:
1단계 : Button 테스트 코드 추가

- 선택된 버튼의 이름이 변경되는 테스트 코드 추가

seong-dodo committed:
2단계: Button 컴포넌트 onClick 기능 구현

- 선택한버튼이 체크되어지도록 기능 구현

seong-dodo committed:
button 관련 오타 및 오류 수정

- button 테스트 오타 수정
- button 테스트 코드 수정
- button 컴포넌트 오류 발견 수정

seong-dodo committed:
Regions 컴포넌트 Categories 컴포넌트 삭제

- buttons 컴포넌트에 합치기

moonkii commented on src/Button.jsx:
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+
+export default function Button({ buttons, onclick, selectedButton }) {
```
버튼들을 구현해주는데 이름이 `Button` 이라 어색한 것 같아요.
복수형태를 해주거나 다른 이름을 해주면 좋을 것 같아요.

moonkii commented on src/Button.test.jsx:
```diff
@@ -0,0 +1,59 @@
+    const handleClick = jest.fn();
+
+    const { queryByText } = render(
+      <Button
+        item={item}
+        onclick={handleClick}
+        selectedButton={beforeSelectedItem}
+      />,
```
구현부랑 테스트랑 인터페이스가 다른데 어떤게 맞나요?

seong-dodo committed:
CategoriesContainer 테스트 코드 추가(코드구현 후 반영)

- handleClick에 대한 테스트 코드 추가

seong-dodo committed:
Regions 컴포넌트 Categories 컴포넌트 삭제전으로 되돌린다

seong-dodo committed:
중간 점검

- 놓친 부분 및 오류 발생 해결 완료 (1차)

seong-dodo committed:
RestaurantsContainer 테스트 코드 추가

- RestaurantsContainer가 그려진다에 대한 테스트 코드 추가

seong-dodo committed:
RestaurantsContainer 구현

- RestaurantsContainer 뼈대 만들기

seong-dodo committed:
reducer 테스트 코드 추가

- 레스토랑 목록 얻기 위한 테스트 코드 추가

seong-dodo committed:
RestaurantsContainer 테스트 코드 추가

- RestaurantsContainer 테스트 리덕스 모킹 추가

seong-dodo committed:
RestaurantsContainer 컴포넌트 구현 수정

- RestaurantsContainer 컴포넌트 리덕스에 상태 가져오도록 구현 수정

seong-dodo committed:
App 테스트 코드 추가

- 레스토랑 목록에 대한 모킹 추가

seong-dodo committed:
레스토랑 목록 가져오기

- API에서 레스토랑 목록 가져오기 기능 구현

moonkii commented on src/Categories.jsx:
```diff
@@ -0,0 +1,35 @@
+  }));
+
+  function handleClick(category) {
+    dispatch(selecteCategory(category));
```
action 을 dispatch 하는 부분도 컨테이너에서 처리해주면 좋을 것 같아요 ;)

moonkii commented on src/Categories.jsx:
```diff
@@ -0,0 +1,35 @@
+export default function Categories({ categories }) {
+  const dispatch = useDispatch();
+
+  const { selectedCategory } = useSelector((state) => ({
+    selectedCategory: state.selectedCategory,
+  }));
```
CategoriesContainer 와 Categories 는 어떤 차이가 있나요?

moonkii commented on src/services/__mocks__-/api.js:
```diff
@@ -0,0 +1,11 @@
+export async function fetchRegions() {
+  return [];
+}
```
mock 함수는 `jest.fn()` 으로 선언해주셔야 해요.

폴더명도 `__mocks__` 으로 해주셔야해요!

seong-dodo replied:
감사합니다 😃

seong-dodo committed:
Button 컴포넌트 테스트 추가 & 줄바꿈 등 코드 수정

- Button 컴포넌트에 loadRestaurants 모킹 함수 추가
- 줄바꿈 및 코드 오류 및 오타 수정

seong-dodo committed:
async action 테스트 작성

moonkii reviewed (approved):
지금처럼 TDD 싸이클을 잘 지켜가면서 꾸준히 하시다보면 테스트도 익숙해 지실거에요!
PR을 코드숨 브랜치 main 으로 보내셨어요. `seong-dodo:main` 으로 PR 보내주시면 merge 하도록하겠습니다!

seong-dodo commented:
감사합니다!!! 제가 아주 큰 실수를 했네요 ㅜㅜ
방금 코드숨 브랜치를 바꿨습니다!!

5주차 과제를 진행할 때 제가 개념을 완벽하게 숙지못해서 일어났던 오류와 이렇게 하지 않아야 할 실수로 인해 많은 시행착오가 생겼는데 이렇게 알려주셔서 감사합니다 🙏

seong-dodo committed:
api에 대한 테스트 코드를 작성한다

- 6주차 과제 참조하여 api에 대한 테스트 코드 작성 완료

seong-dodo committed:
피드백 반영

- 줄바꿈 및 주석 제거
- getBy~와 queryBy~ 에 대한 고민 후 수정
- 테스트 시나리오 네이밍 점검
